Add Hide eCommerce Experiment from Launch Site Flow

dzver opened the pull request:
#### Changes proposed in this Pull Request

Hides the eCommerce plan for 50% of the users in the launch-site flow

<img width="1185" alt="Screenshot 2021-04-15 at 16 37 26" src="https://user-images.githubusercontent.com/82778/114885202-1b9e7680-9e0f-11eb-9cd8-172bb87cc345.png">

I previously prepared #51068 but the new ExPlat client was introduced when I worked on it plus several other changes happened, including the PlanFeaturesComparison.

#### Testing instructions

Make sure to check each step on mobile resolutions!

1. Assign yourself to the test variation of the `hide_ecommerce_launch_site ` in ExPlat
2. Create a new user and site. Make sure that on the Signup Plans step, the eCommerce plan is visible
3. Go to /plans - eCommerce should be visible
3. Launch the site, stop at the plans step
4. Make sure that on the Plans step, the eCommerce plan is not visible
5. Switch between Annual and Monthly plans. It should keep the eCommerce hidden
6. Switch to Control and reload
7. No eCommerce should now be visible

There's some duplication but it is because the code I change is duplicated. I don't think additional helpers are needed given that it's an experiment.

niranjan-uma-shankar commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -551,13 +549,38 @@ export class PlansFeaturesMain extends Component {
+		}
+
+		return (
+			<ProvideExperimentData name="hide_ecommerce_launch_site">
```
I recommend adding the new `isEligible` prop to check if it's the launch page - check this [review comment](https://github.com/Automattic/wp-calypso/pull/50512/files#r615207493) on a different PR on how it works.

niranjan-uma-shankar commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -952,9 +954,16 @@ const ConnectedPlanFeatures = connect(
 			! isInSignup && ! isJetpack && kindOfPlanTypeSelector === 'interval';
+		const isInHideECommerceExperiment =
+			isInSignup && isLaunchPage && 'treatment' === experimentVariation;
+		const filteredPlans = isInHideECommerceExperiment
```
I suggest filtering the plans in the function responsible for it - `getPlansForPlanFeatures()` [here](https://github.com/Automattic/wp-calypso/blob/05732f098d92569f4fb9d8398ade6f9c2006e339/client/my-sites/plans-features-main/index.jsx#L285) so that we can keep related logic in the same place.

dzver replied:
I had to shuffle around the ProvideExperimentData in order to make this work.

getPlansForPlanFeatures is called twice in the same render(), it's not clear why.

niranjan-uma-shankar commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -952,9 +954,16 @@ const ConnectedPlanFeatures = connect(
 		const isLoggedInMonthlyPricing =
 			! isInSignup && ! isJetpack && kindOfPlanTypeSelector === 'interval';
+		const isInHideECommerceExperiment =
+			isInSignup && isLaunchPage && 'treatment' === experimentVariation;
```
I think `'treatment' === experimentVariation` should suffice since treatment is assigned only when the previous two conditions are met?

niranjan-uma-shankar reviewed (commented):
When there are only three plans visible, the plan boxes expand to fit the available width and as a result of the wider boxes, there's more whitespace which makes the UI a bit off. IMO it would look better if we can reduce the width a bit.

<img width="1289" alt="Screenshot 2021-04-19 at 9 54 08 AM" src="https://user-images.githubusercontent.com/1269602/115188218-3bc98080-a102-11eb-9bf6-314a4807a299.png">

dzver committed:
Add hideEcommercePlan experiment

I'm not adding it to the parent component to prevent conflict with another experiment

The calls to getPlansForPlanFeatures are already duplicate before this
PR. Deduplication can be addressed separately but is not important
given the overall trend to migrate to gutenboarding signup.

dzver committed:
Make the comparison table smaller

niranjan-uma-shankar commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -216,7 +217,7 @@ export class PlansFeaturesMain extends Component {
 		} = this.props;
 
-		const plans = this.getPlansForPlanFeatures();
+		const plans = this.getPlansForPlanFeatures( experimentVariation );
```
I think this change is not needed since `getPlanFeatures()` is used to render the /plans page in Calypso and is not used for the launch flow.

dzver replied:
Unfortunately, if I remove this change, the experiment will no longer apply on mobile and eCommerce will be visible.

niranjan-uma-shankar replied:
Ah yes, you're right. My bad :)

niranjan-uma-shankar commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -129,7 +130,7 @@ export class PlansFeaturesMain extends Component {
 	}
 
-	showFeatureComparison() {
+	showFeatureComparison( experimentVariation = null ) {
```
Nitpick: The `null` assignment is not needed since that's the default value.

niranjan-uma-shankar commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -274,7 +275,7 @@ export class PlansFeaturesMain extends Component {
 	}
 
-	getPlansForPlanFeatures() {
+	getPlansForPlanFeatures( experimentVariation = null ) {
```
Same as before, `null` is not needed as that's the default value.

dzver replied:
Thanks!

niranjan-uma-shankar reviewed (commented):
Thanks for working through these changes. 

Can we add a loading state in the UI when the plans step is waiting to render while fetching the experiment data? (notice the brief blank section in the gif below). 
![2204](https://user-images.githubusercontent.com/1269602/115712647-a42e9100-a392-11eb-9069-e7420d2ea25c.gif) 

We have a pulsing dot but the `isLoadingExperiment` needs to be defined for this experiment:

https://github.com/Automattic/wp-calypso/blob/9e69a131fbed1a6ad4d017bfefcbdd2919541785/client/signup/steps/plans/index.jsx#L171-L174

As an addendum to the previous point, @jessie-ross also recommends [here](https://github.com/Automattic/wp-calypso/pull/50512#discussion_r604806345) to add `loadExperimentAssignment` to reduce the chance of a loading state.

dzver committed:
Address feedback by Niranjan

- default values removed from a few places where that's unnecessary
- Pulsing dot added for when the experiment is loading
- Experiment data is loaded early so that the white pulse dot is never
  visible

jessie-ross commented on client/signup/controller.js:
```diff
@@ -100,6 +100,7 @@ export default {
 	redirectTests( context, next ) {
 		const currentFlowName = getFlowName( context.params );
 		currentFlowName === 'onboarding' && loadExperimentAssignment( 'refined_reskin_v1' );
+		currentFlowName === 'launch-site' && loadExperimentAssignment( 'hide_ecommerce_launch_site' );
```
Hmm, could be placed a bit later than on load, like at the signup step? 
Ideally it is placed at "the UI just before the UI you are testing".

For example the `refined_reskin_v1` is here as it is just before the signup screen.

jessie-ross replied:
It looks like the eligibility here (`currentFlowName === 'launch-site'`) doesn't match the eligibility in `ProvideExperimentData` (`isLaunchPage && isInSignup`), ideally they are the same.

dzver replied:
Thanks for the suggestion.

There is another flow 'new-launch' defined that could have the same isLaunchPage = true. Let's make sure it doesn't happen, although, I don't think the gutenboarding launch flow uses any of this code.

dzver commented:
I tried running the e2e tests on trunk and saw the same errors.

dzver committed:
Make the eligibility check identical

niranjan-uma-shankar reviewed (approved):
LGTM 🚢
